doc/rpldev.4
.TH "rpldev" "4" "2009-09-27" "ttyrpld" "tty logging daemon suite"
.SH "Name"
.PP
rpldev \(em tty logger kernel module
.SH Description
.PP
\fB/dev/rpl\fP is a character device file which returns the data
captured from ttys through the rpldev kernel module. It has no fixed
device number, since it asks for a dynamic major/minor number. The
Kernel module will emit a notice where it actually landed. The kernel
hotplug framework and udev will then create a device node in /dev.
.PP
On FreeBSD, devfs takes care of all this, so when the module is
loaded, a working /dev/rpl is also present. For OpenBSD and NetBSD,
it always loads at char(228,0). Use `\fBmknod /dev/rpl c 228 0\fP` to
create the node. It should be made mode 0400 and owned by the
low-privileged user that is set in rpld.conf.
.SH "Module parameters (Linux)"
.PP
The rpldev module has the following options. Some of them may be
changed dynamically at run-time via /sys/module/rpldev/.
.TP
Bufsize
Size of the ring buffer in bytes. Default is 32K.
.TP
Minor_nr
Force a minor number rather than automatically deciding for one. Passing 255
here will use auto-selection, and is the default.
.PP
These options can be provided to modprobe like this:
.PP
.nf
	# modprobe rpldev Minor_nr=37
.fi
.PP
You can put the options into /etc/modprobe.conf.local, to be
automatically sourced when modprobe is called:
.PP
.nf
	options rpldev Minor_nr=37
.fi
.PP
I have not found any way for passing or modifying options under BSD,
so unless you are willing to recompile the kernel, you are stuck with
a fixed 32K buffer.
.SH "Data structures"
.PP
Everytime the kernel functions for reading/writing from/to a terminal
are executed, the call chain is extened into the rpldev hooks, i.e.
an external function in the rpldev kernel module is called, which
transfers the tty buffer into so-called packets, which are then
returned as /dev/rpl is read.
.PP
A packet consists of the event type (which also contains a magic number
for file identification), device number, size
of the data, timestamp, and of course, the data itself. All fields
are little-endian and packed, i.e. there are no alignment gaps.
When using \fIevmagic\fP however, it must be interpreted as big-endian.
The purpose is to have the magic string "rp3@". The struct
can be represented in this C struct:
.PP
.nf
struct rpltime {
	uint64_t tv_sec;
	uint32_t tv_usec;
};

union rpldev_evmagic {
	uint32_t evmagic; /*BE*/
	char magic[4];
};

struct rpldev_packet {
	union rpldev_evmagic evmagic;
	uint32_t size; /*LE*/
	struct rpltime time; /*LE*/
	uint32_t dev; /*LE*/
} __attribute__((packed));
.fi
.PP
The data itself is actually not included in the struct, since it is of dynamic
length. For further reference, the struct alone will be called packet header,
whereas packet header plus data (or payload) equals packet. There are no
byte-aligning gaps between the members, i.e. it is a so-called packed
structure.
.PP
The \fIdev\fP member contains the device number on which the event occurred.
Since both the master and the slave side of a pty pair can generate events, and
we usually do not need the master side events, packets from the master sides
are already dropped at the kernel level.
.PP
The device field is made up of 12 bits for the major number and 20 bits for the
minor number, in little-endian, and has this very same layout on all rpldev
architectures. (For reference, this 12/20 split comes from Linux 2.6.) BSD
rpldev will convert device numbers to this scheme. As a result, a truncation is
involved, since BSD uses an 8/24 split, so minors greater than 1048575 will be
truncated. This should not pose a problem, as one million devices is already
more than a system could ever have in a meaningful way.
.PP
The following event types exist.
.IP \(bu 4
RPLEVT_NONE = 0x7270334E, /* "rp3@" */
.IP
RPLEVT_OPEN = 0x72703341, /* "rp3A" */
.IP \(bu 4
RPLEVT_READ = 0x72703342, /* "rp3B" */
.IP \(bu 4
RPLEVT_WRITE = 0x72703343, /* "rp3C" */
.IP \(bu 4
RPLEVT_LCLOSE = 0x72703344, /* "rp3D" */
.PP
RPLEVT_OPEN, RPLEVT_READ and RPLEVT_WRITE are generated whenever an open(),
read() or write() is done, respectively. RPLEVT_LCLOSE is generated when the
last file descriptor to the tty is closed.
.PP
RPLEVT_NONE are no-ops. The kernel modules will not output this type, but
rpld(8) will add it at the start of a logfile to ease identification by
file(1). rpld outputs the ident by setting \fIsize\fP to 0 (because there is
no data following) and the time field is filled with 0xFA.
.PP
During an open() on a tty line (Linux only), the filename of the accessed
dentry is extracted and passed to rpld within the payload, for convenience, so
that \fBrpld\fP(8) does not need to do a time-consuming search for it in /dev.


doc/changelog.txt

Notes
=====

You do not need to repatch your kernel or recompile rpldev if they have not
changed significantly between ttyrpld releases.

│                │last time kpatch│                            │
│Operating system│(rpldhk) changed│last time rpldev changed:   │
├────────────────┼────────────────┼────────────────────────────┤
│Linux           │2.50            │2.60                        │
│FreeBSD         │2.60            │2.60                        │
│OpenBSD         │-               │2.60                        │
│NetBSD          │2.50            │2.60                        │


Changelog
=========

HEAD

Changes:
- doc: move changelog to text/plain
- rpldev: NetBSD 5.0.1 support
- rpldev: OpenBSD 4.5 support
- rpldev: Linux 2.6.31 support
- rpldev: FreeBSD 8.0 support
- packet: increase the "size" field to 32 bits
- packet: combine "magic" and "event" fields, and move to front


v2.52 (2008-09-07)

  • Updated kernel components for Linux 2.6.27
  • Updated userspace code for libHX 1.25

v2.51 (2008-05-15)

  • Updated rpldhk and rpldev for Linux 2.6.25, OpenBSD 4.3, FreeBSD 7.0,
    NetBSD 4.0

v2.50 (2008-01-01)

  • changed the packet format to be compatible between 32-bit and 64-bit
  • added rplcvt utility for converting old on-disk rpld files
  • explicitly blacklist /dev/stdout, /dev/stderr and /dev/stderr - these have
    been popping up as tty names because there is such a symlink in Linux's /
    dev. (But we need symlink processing for Solaris, eww.)

v2.19 (2007-09-17)

  • The kernel interface has been reduced (hooks removed). The 2.19 userspace
    is still compatible (throws warnings though) with 2.18 kernel components
    however. The 2.19 kernel patch is for 2.6.22.
  • an experimental Solaris kernel driver has been added - tested with Solaris
    11.snv only, but may work with older ones.
  • fix a file descriptor leak in rpld (thanks J.Eller)
  • fix a symlink recursion when walking /dev (J.Eller)
  • raise maximum number of open files (J.Eller)
  • handle permission restrictions on /dev nodes
  • fix directory name of Persian translation
  • moved to autotools
  • build with new libHX 1.10

v2.18 (2006-07-01)

  • Fixed OpenBSD 3.9 kpatch
  • Readded workaround for OpenBSD/NetBSD EINTR signal handling that got lost
    in 2.17

v2.17 (2006-06-19)

  • Support for Linux 2.6.17, FreeBSD 6.1, OpenBSD 3.9
  • added missing -fprofile-arcs -ftest-coverage to LDFLAGS for PROF=1

v2.16 (2006-04-01)

  • user/rpld.c: removed a superfluous variable
  • Added polling functions to the BSD rpldevs. This fixes the high CPU usage
    (which is due to the implementation of pthreads on OpenBSD) observed with
    rpld/OpenBSD.
  • OpenBSD part is again just a patch since modules are said to be unsupported
    on non-i386
  • Due to popular request, the docs now come with a black-on-white ("paper
    color") color scheme and no fixed font size. The old gray-on-black
    ("console color") is available as an alternate stylesheet and can be used
    in Firefox using View > Page Style > Black background.
  • Added the EVT_MAGIC pseudo event type
  • Added share/ttyrpld.magic for the file(1) utility

v2.15 (2006-03-11)

General:

  • added support for NetBSD 3.0
  • shifts for majors>255 and minors>255 were broken in include/dev.h

kpatch:

  • [All] fixed some SMP race conditions
  • [OpenBSD] fixed: rpl_deinit() was not called in ttyclose()

rpldev:

  • [All] rename functions to make function more clear (mv_to_user ->
    circular_get, mv_buffer -> circular_put_packet, mv_buffer2 -> circular_put)
  • [All] make avail_R() and avail_W() return size_t instead of ssize_t,
    addressing some potential overflow problems
  • [Linux] document circulary calculation of avail_*(), please report errors
  • [FreeBSD] merged freebsd-*/rpldev.c into one file
  • [FreeBSD/OpenBSD] removed SKIP_PTM, since master ptys do not generate any
    noise (that's a kpatch feature)
  • [OpenBSD] rpldev is now a module (rpldev_mod.o)
  • [OpenBSD] loads at new device number (228,0), please re-mknod your /dev/rpl
    node

v2.12 (2006-01-21)

  • fixed: FreeBSD 6.0 hands krpl_close() a NULL pointer in some cases; handle
    it properly
  • some doc updates
  • GNUmakefile: "all" target does not include "doc" anymore, because BSDs
    (standard installation) often ship without php, w3m and perl.
  • added the '[' and ']' keys to ttyreplay (lower/raise speed by 0.10)

v2.11 (2006-01-06)

Added support back in for the latest releases of the previous generation of
kernels. For Linux, that's 2.4.something (USE DISCOURAGED!), and for FreeBSD
that's 5.3.

v2.10 (2005-12-04)

Support for OpenBSD 3.8 and FreeBSD 6.0 was added; permission denied, syslog
spamming, cpu hog, ownership and symlink recursion bugs were fixed. Docs were
updated to reflect advances since the last seven releases.

  • Support for OpenBSD 3.8 and FreeBSD 6.0
  • Documentation has been updated to reflect advances since the last seven
    releases.
  • user/replay.c: fixed a color bug when -t was active (spotted whilst under
    FreeBSD)
  • user/*.c: use usleep(10000) instead of sched_yield() to reduce idle cpu hog
  • user/rpld.c: umask was set wrong, so 1.) logging did not take place 2.)
    syslog was spammed with permission denied warnings
  • user/rpld.c: added a rate limiter for syslog messages
  • user/rpld.c: find_devnode_dive() forgot to check for symlinks so that we
    dived into /dev/fd, which is not wanted, because /dev/fd/* itself includes
    /dev again, which lead to unwanted results.
  • user/rpld.c: fixed a filedescriptor leak in fill_info()
  • user/rpld.c: another try at getting ownerships right; rpld still had a case
    where the recorded owner differed from the owner of the /dev node (with
    respect to the design decision). See one below:
  • linux-2.6/rpldev.c: /dev/tty is a special case with respect to ownerships,
    so it must be filtered out in krpl_init() and krpl_open().

doc/rpldev.4.php
<?php include_once("_header.php"); ?>

<h1>rpldev - tty logger kernel module</h1>

<p class="block"><code>/dev/rpl</code> is a character device file which returns
the data captured from ttys through the <i>rpldev</i> kernel module. It has no
fixed device number, since it asks for a dynamic major/minor number. Usually,
it will get <code>char(10,63)</code> or <code>char(10,62)</code> on
Linux&nbsp;-- the Kernel module will emit a notice where it actually landed.
The kernel hotplug framework and <i>udev</i> will then create a device node in
<code>/dev</code>.</p>

<p class="block">On FreeBSD, devfs takes care of all this, so when the module
is loaded, a working <code>/dev/rpl</code> is also present. For OpenBSD and
NetBSD, it always loads at <code>char(228,0)</code>. Use <code>mknod /dev/rpl c
228 0</code> to create the node. It should be made mode <code>0400</code> and
owned by the low-privileged user that is set in <code>rpld.conf</code>.</p>

<h1>Module parameters (Linux)</h1>

<p class="block">The <i>rpldev</i> module has the following options. Some of
them may be changed dynamically at run-time via
<code>/sys/module/rpldev/</code>.</p>

<table border="1">
  <tr>
    <td class="t1">Bufsize</td>
    <td class="t1">Size of the ring buffer in bytes. Default is 32K.</td>
  </tr>
  <tr>
    <td class="t2">Minor_nr</td>
    <td class="t2">Force a minor number rather than automatically deciding for
      one.  Passing 255 here will use auto-selection, and is the default.</td>
  </tr>
</table>

<p class="block">These options can be provided to <i>modprobe</i> like
this:</p>

<p class="code"><code><b>#</b> modprobe rpldev Minor_nr=37</code></p>

<p class="block">You can put the options into
<code>/etc/modprobe.conf.local</code>, to be automatically sourced when
modprobe is called:</p>

<p class="code"><code>options rpldev Minor_nr=37</code></p>

<p class="block">I have not found any way for passing or modifying options
under BSD, so unless you are willing to recompile the kernel, you are stuck
with a fixed 32K buffer.</p>

<h1>Data structures</h1>

<p class="block">Everytime the kernel functions for reading/writing from/to a
terminal are executed, the call chain is extened into the rpldev hooks, i.e. an
external function in the <i>rpldev</i> kernel module is called, which transfers
the tty buffer into so-called packets, which are then returned as
<code>/dev/rpl</code> is read.</p>

<p class="block">A packet consists of a device number, magic byte (always
<code>0xEE</code>), size of the data, timestamp, and of course, the data
itself. All fields are little-endian and packed, i.e. there are no alignment
gaps. It can be represented in this C struct:</p>

<p class="code"><code>
<b>struct</b> rpltime {<br />
&nbsp; &nbsp; <b>uint32_t</b> tv_sec;<br />
&nbsp; &nbsp; <b>uint64_t</b> tv_usec;<br />
};<br />
<br />
<b>struct</b> rpldev_packet {<br />
&nbsp; &nbsp; <b>uint32_t</b> dev;<br />
&nbsp; &nbsp; <b>uint16_t</b> size;<br />
&nbsp; &nbsp; <b>uint8_t</b> event, magic;<br />
&nbsp; &nbsp; <b>struct</b> rpltime time;<br />
} <i>__attribute__((packed))</i>;</code></p>

<p class="block">The data itself is actually not included in the struct, since
it is of dynamic length. For further reference, the struct alone will be called
<i>packet header</i>, whereas packet header plus <i>data</i> (or
<i>payload</i>) equals <i>packet</i>. There are no byte-aligning gaps between
the members, i.e. it is a so-called <i>packed</i> structure. (No pun between
packet and packed.)</p>

<p class="block">The <code>-&gt;dev</code> member contains the device number on
which the event occurred. Since both the master and the slave side of a pty
pair can generate events, and we usually do not need the master side events,
packets from the master sides are already dropped at the kernel level.</p>

<p class="block">The device field is made up of 12 bits for the major number
and 20 bits for the minor number, in little-endian, and has this very same
layout on all rpldev architectures. (For reference, this 12/20 split comes from
Linux 2.6.) BSD rpldev will convert device numbers to this scheme. As a result,
a truncation is involved, since BSD uses an 8/24 split, so minors greater than
1048575 will be truncated. This should not pose a problem, as one million
devices is already more than a system could ever have in a meaningful way.</p>

<p class="block">The following event types exist:</p>

<p class="code"><code>
<b>enum</b> {<br />
&nbsp; &nbsp; EVT_OPEN <b>=</b> 0x01,<br />
&nbsp; &nbsp; EVT_READ <b>=</b> 0x02,<br />
&nbsp; &nbsp; EVT_WRITE <b>=</b> 0x03,<br />
&nbsp; &nbsp; EVT_LCLOSE <b>=</b> 0x64,<br />
};</code></p>

<p class="block"><code>EVT_OPEN</code>, <code>EVT_READ</code> and
<code>EVT_WRITE</code> are generated whenever an <code>open()</code>,
<code>read()</code> or <code>write()</code> is done, respectively.
<code>EVT_LCLOSE</code> is generated when the last file descriptor to the tty
is closed.</p>

<p class="block">During an <code>open()</code> on a tty line (Linux only), the
filename of the accessed dentry is extracted and passed to <i>rpld</i> within
the payload, for convenience, so that <i>rpld</i> does not need to do a
time-consuming search for it in <code>/dev</code>.</p>

<?php include_once("_footer.php"); ?>
